idx/memory/tag_query.go

```diff
@@ -0,0 +1,180 @@
+package memory
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/raintank/metrictank/idx"
+)
+
+var (
+	expressionRe    = regexp.MustCompile("^([^!=~]+)([!=~]{1,3})(.*)$")
+	errUnknownId    = errors.New("id does not exist")
+	errInvalidTag   = errors.New("found an invalid tag in the index")
+	errInvalidQuery = errors.New("invalid query")
+)
+
+type TagQuery struct {
+	selects []func(TagIndex) (map[string]struct{}, error)
+	filters []func(def *idx.Archive) (bool, error)
+}
+
+func NewTagQuery(expressions []string) (*TagQuery, error) {
+	if len(expressions) == 0 {
+		return nil, errInvalidQuery
+	}
+
+	query := &TagQuery{}
+
+	for _, expr := range expressions {
+		if !expressionRe.MatchString(expr) {
+			return nil, errInvalidQuery
+		}
+
+		exprSplits := expressionRe.FindAllStringSubmatch(expr, -1)
+		if len(exprSplits[0]) != 4 {
+			return nil, errInvalidQuery
+		}
+
+		key := exprSplits[0][1]
+		operator := exprSplits[0][2]
+		value := exprSplits[0][3]
+
+		// always anchor all regular expressions at the beginning
+		if operator[len(operator)-1] == byte('~') && len(value) > 0 && value[len(value)-1] != byte('^') {
+			value = fmt.Sprintf("^%s", value)
+		}
+
+		// special case of empty value
+		if len(value) == 0 {
+		}
```

Comment:
this is a no-op

Reply:
yeah, still going to add that. the idea is that i'll just reverse the operation

Reply:
then, further down where the matching is done, i could also shortcut it to not even do the mapping if the pattern is `.*` or `.+`. (for `.+` this works assuming that no value that's in the index can be `""`)